File: android/app/src/main/kotlin/com/turqapp/app/qa/PlaybackHealthMonitor.kt
package com.turqapp.app.qa

import android.os.SystemClock
import android.util.Log
import java.util.LinkedHashSet

/**
 * Collects feed / shorts playback truth and deduplicated critical errors.
 */
class PlaybackHealthMonitor(
    private val tag: String = "PlaybackHealthMonitor",
    private val firstFrameTimeoutMs: Long = 1_500L,
    private val freezeFrameTimeoutMs: Long = 1_000L,
    private val excessiveDroppedFramesThreshold: Int = 24,
    private val excessiveRebufferThreshold: Int = 3,
) {
    var stateListener: ((PlaybackHealthMonitor) -> Unit)? = null

    private val errorLock = Any()
    private val recordedErrors = LinkedHashSet<String>()

    @Volatile
    var playbackRequestedAt: Long = 0L
        private set

    @Volatile
    var playerReadyAt: Long = 0L
        private set

    @Volatile
    var firstFrameRenderedAt: Long = 0L
        private set

    @Volatile
    var lastFrameRenderedAt: Long = 0L
        private set

    @Volatile
    var lastKnownPlaybackPosition: Long = 0L
        private set

    @Volatile
    var isPlaybackExpected: Boolean = false
        private set

    @Volatile
    var isPlaying: Boolean = false
        private set

    @Volatile
    var hasRenderedFirstFrame: Boolean = false
        private set

    @Volatile
    var isBuffering: Boolean = false
        private set

    @Volatile
    var isInFullscreenTransition: Boolean = false
        private set

    @Volatile
    var surfaceAttachCount: Int = 0
        private set

    @Volatile
    var surfaceDetachCount: Int = 0
        private set

    @Volatile
    var droppedFramesTotal: Int = 0
        private set

    @Volatile
    var rebufferCount: Int = 0
        private set

    @Volatile
    var lastKnownPlaybackAdvanceAt: Long = 0L
        private set

    @Volatile
    var fullscreenTransitionStartedAt: Long = 0L
        private set

    @Volatile
    var appBackgroundedAt: Long = 0L
        private set

    @Volatile
    var appForegroundedAt: Long = 0L
        private set

    @Volatile
    var awaitingFullscreenRecovery: Boolean = false
        private set

    @Volatile
    var awaitingBackgroundRecovery: Boolean = false
        private set

    @Volatile
    var playerReadyObserved: Boolean = false
        private set

    fun resetForNewPlaybackSession() {
        playbackRequestedAt = 0L
        playerReadyAt = 0L
        firstFrameRenderedAt = 0L
        lastFrameRenderedAt = 0L
        lastKnownPlaybackPosition = 0L
        isPlaybackExpected = false
        isPlaying = false
        hasRenderedFirstFrame = false
        isBuffering = false
        isInFullscreenTransition = false
        surfaceAttachCount = 0
        surfaceDetachCount = 0
        droppedFramesTotal = 0
        rebufferCount = 0
        lastKnownPlaybackAdvanceAt = 0L
        fullscreenTransitionStartedAt = 0L
        appBackgroundedAt = 0L
        appForegroundedAt = 0L
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        playerReadyObserved = false
        synchronized(errorLock) {
            recordedErrors.clear()
        }
        publish("reset")
    }

    fun onPlaybackRequested() {
        val timestamp = now()
        playbackRequestedAt = timestamp
        playerReadyAt = 0L
        firstFrameRenderedAt = 0L
        lastFrameRenderedAt = 0L
        lastKnownPlaybackPosition = 0L
        lastKnownPlaybackAdvanceAt = 0L
        isPlaybackExpected = true
        isPlaying = false
        hasRenderedFirstFrame = false
        isBuffering = false
        playerReadyObserved = false
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        log("playbackRequested at=$timestamp")
        publish("playbackRequested")
    }

    fun onPlayerReady() {
        if (!playerReadyObserved) {
            playerReadyObserved = true
            playerReadyAt = now()
        }
        log("playerReady at=$playerReadyAt")
        evaluatePassiveTimeouts()
        publish("playerReady")
    }

    fun onPlaybackStarted() {
        isPlaying = true
        isPlaybackExpected = true
        isBuffering = false
        clearErrors("PLAYBACK_NOT_STARTED")
        publish("playbackStarted")
    }

    fun onPlaybackPaused() {
        isPlaying = false
        publish("playbackPaused")
    }

    fun onFirstFrameRendered() {
        val timestamp = now()
        hasRenderedFirstFrame = true
        if (firstFrameRenderedAt == 0L) {
            firstFrameRenderedAt = timestamp
        }
        lastFrameRenderedAt = timestamp
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        clearErrors(
            "FIRST_FRAME_TIMEOUT",
            "READY_WITHOUT_FRAME",
            "PLAYBACK_NOT_STARTED",
        )
        log("firstFrameRendered ttffMs=${deltaFrom(playbackRequestedAt, timestamp)}")
        publish("firstFrameRendered")
    }

    fun onFrameRendered() {
        lastFrameRenderedAt = now()
        if (!hasRenderedFirstFrame) {
            hasRenderedFirstFrame = true
            if (firstFrameRenderedAt == 0L) {
                firstFrameRenderedAt = lastFrameRenderedAt
            }
        }
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        publish("frameRendered")
    }

    fun onPositionUpdate(positionMs: Long) {
        if (positionMs > lastKnownPlaybackPosition + 40L) {
            lastKnownPlaybackAdvanceAt = now()
        }
        lastKnownPlaybackPosition = positionMs
        publish("position=$positionMs")
    }

    fun onBufferingStarted() {
        if (!isBuffering) {
            rebufferCount += 1
            if (rebufferCount >= excessiveRebufferThreshold) {
                addError("EXCESSIVE_REBUFFERING")
            }
        }
        isBuffering = true
        publish("bufferingStarted")
    }

    fun onBufferingEnded() {
        isBuffering = false
        publish("bufferingEnded")
    }

    fun onAudioMissing() {
        addError("AUDIO_NOT_STARTED")
    }

    fun onPlaybackNotStarted() {
        addError("PLAYBACK_NOT_STARTED")
    }

    fun onFullscreenTransitionStarted() {
        fullscreenTransitionStartedAt = now()
        isInFullscreenTransition = true
        awaitingFullscreenRecovery = isPlaybackExpected || isPlaying || hasRenderedFirstFrame
        publish("fullscreenTransitionStarted")
    }

    fun onFullscreenTransitionEnded() {
        isInFullscreenTransition = false
        if (awaitingFullscreenRecovery) {
            fullscreenTransitionStartedAt = now()
        }
        publish("fullscreenTransitionEnded")
    }

    fun onSurfaceAttached() {
        surfaceAttachCount += 1
        if (!hasRenderedFirstFrame && surfaceAttachCount >= 2) {
            addError("DOUBLE_BLACK_SCREEN_RISK")
        }
        publish("surfaceAttached")
    }

    fun onSurfaceDetached() {
        surfaceDetachCount += 1
        publish("surfaceDetached")
    }

    fun onAppBackgrounded() {
        appBackgroundedAt = now()
        awaitingBackgroundRecovery = isPlaybackExpected || isPlaying || hasRenderedFirstFrame
        isPlaying = false
        publish("appBackgrounded")
    }

    fun onAppForegrounded() {
        appForegroundedAt = now()
        publish("appForegrounded")
    }

    fun onDroppedFrames(count: Int) {
        droppedFramesTotal += count
        if (droppedFramesTotal >= excessiveDroppedFramesThreshold) {
            addError("EXCESSIVE_DROPPED_FRAMES")
        }
        publish("droppedFrames")
    }

    fun onVideoFreeze() {
        addError("VIDEO_FREEZE")
    }

    fun onReadyWithoutFrame() {
        addError("READY_WITHOUT_FRAME")
    }

    fun onFirstFrameTimeout() {
        addError("FIRST_FRAME_TIMEOUT")
    }

    fun onFullscreenInterruption() {
        addError("FULLSCREEN_INTERRUPTION")
    }

    fun onBackgroundResumeFailure() {
        addError("BACKGROUND_RESUME_FAILURE")
    }

    fun getErrors(): List<String> {
        return synchronized(errorLock) { recordedErrors.toList() }
    }

    fun hasErrors(): Boolean {
        return synchronized(errorLock) { recordedErrors.isNotEmpty() }
    }

    fun hasFreshFrameSince(timestamp: Long): Boolean {
        return lastFrameRenderedAt >= timestamp && lastFrameRenderedAt > 0L
    }

    fun snapshot(): Map<String, Any> = mapOf(
        "playbackRequestedAt" to playbackRequestedAt,
        "playerReadyAt" to playerReadyAt,
        "firstFrameRenderedAt" to firstFrameRenderedAt,
        "lastFrameRenderedAt" to lastFrameRenderedAt,
        "lastKnownPlaybackPosition" to lastKnownPlaybackPosition,
        "isPlaybackExpected" to isPlaybackExpected,
        "isPlaying" to isPlaying,
        "hasRenderedFirstFrame" to hasRenderedFirstFrame,
        "isBuffering" to isBuffering,
        "isInFullscreenTransition" to isInFullscreenTransition,
        "surfaceAttachCount" to surfaceAttachCount,
        "surfaceDetachCount" to surfaceDetachCount,
        "droppedFramesTotal" to droppedFramesTotal,
        "rebufferCount" to rebufferCount,
        "appBackgroundedAt" to appBackgroundedAt,
        "appForegroundedAt" to appForegroundedAt,
        "awaitingFullscreenRecovery" to awaitingFullscreenRecovery,
        "awaitingBackgroundRecovery" to awaitingBackgroundRecovery,
        "errors" to getErrors(),
    )

    fun evaluatePassiveTimeouts() {
        val timestamp = now()
        if (playbackRequestedAt > 0L &&
            !hasRenderedFirstFrame &&
            timestamp - playbackRequestedAt > firstFrameTimeoutMs
        ) {
            onFirstFrameTimeout()
            if (!isPlaying && isPlaybackExpected) {
                onPlaybackNotStarted()
            }
        }

        if (playerReadyAt > 0L &&
            !hasRenderedFirstFrame &&
            timestamp - playerReadyAt > firstFrameTimeoutMs
        ) {
            onReadyWithoutFrame()
        }

        if (awaitingFullscreenRecovery &&
            fullscreenTransitionStartedAt > 0L &&
            timestamp - fullscreenTransitionStartedAt > firstFrameTimeoutMs &&
            isPlaybackExpected &&
            !hasFreshFrameSince(fullscreenTransitionStartedAt)
        ) {
            onFullscreenInterruption()
        }

        if (awaitingBackgroundRecovery &&
            appForegroundedAt > 0L &&
            timestamp - appForegroundedAt > firstFrameTimeoutMs &&
            isPlaybackExpected &&
            !hasFreshFrameSince(appForegroundedAt)
        ) {
            onBackgroundResumeFailure()
        }
    }

    fun shouldFlagVideoFreeze(timestamp: Long = now()): Boolean {
        if (!hasRenderedFirstFrame || lastFrameRenderedAt <= 0L) return false
        if (!(isPlaybackExpected || isPlaying || lastKnownPlaybackAdvanceAt > 0L)) return false
        return timestamp - lastFrameRenderedAt > freezeFrameTimeoutMs
    }

    private fun addError(code: String) {
        val inserted = synchronized(errorLock) { recordedErrors.add(code) }
        if (!inserted) return
        Log.e(tag, "error=$code snapshot=${snapshot()}")
        publish("error=$code")
    }

    private fun clearErrors(vararg codes: String) {
        if (codes.isEmpty()) return
        val changed = synchronized(errorLock) {
            recordedErrors.removeAll(codes.toSet())
        }
        if (!changed) return
        publish("clear=${codes.joinToString(",")}")
    }

    private fun publish(event: String) {
        log(event)
        stateListener?.invoke(this)
    }

    private fun deltaFrom(start: Long, end: Long): Long {
        if (start <= 0L || end <= 0L || end < start) return 0L
        return end - start
    }

    private fun log(message: String) {
        Log.d(tag, message)
    }

    private fun now(): Long = SystemClock.elapsedRealtime()
}

File: ios/Runner/PlaybackHealthMonitor.swift
import Foundation
import QuartzCore

final class PlaybackHealthMonitor {
    typealias StateListener = (PlaybackHealthMonitor) -> Void

    private let tag: String
    private let firstFrameTimeout: CFTimeInterval
    private let freezeFrameThreshold: CFTimeInterval
    private let excessiveStallThreshold: Int

    private var recordedErrors = Set<String>()
    private var errorOrder = [String]()

    var stateListener: StateListener?

    private(set) var playbackRequestedAt: CFTimeInterval = 0
    private(set) var playerReadyAt: CFTimeInterval = 0
    private(set) var firstFrameRenderedAt: CFTimeInterval = 0
    private(set) var lastFrameRenderedAt: CFTimeInterval = 0
    private(set) var lastKnownPlaybackTime: Double = 0

    private(set) var isPlaybackExpected = false
    private(set) var isPlaying = false
    private(set) var hasRenderedFirstFrame = false
    private(set) var isBuffering = false
    private(set) var isInFullscreenTransition = false
    private(set) var isLayerAttached = false

    private(set) var stallCount = 0
    private(set) var layerAttachCount = 0
    private(set) var fullscreenTransitionStartedAt: CFTimeInterval = 0
    private(set) var appDidEnterBackgroundAt: CFTimeInterval = 0
    private(set) var appWillEnterForegroundAt: CFTimeInterval = 0
    private(set) var lastPlaybackProgressedAt: CFTimeInterval = 0
    private(set) var awaitingFullscreenRecovery = false
    private(set) var awaitingBackgroundRecovery = false
    private(set) var lastAccessLogSummary = ""

    init(
        tag: String = "PlaybackHealthMonitor",
        firstFrameTimeout: CFTimeInterval = 1.5,
        freezeFrameThreshold: CFTimeInterval = 1.0,
        excessiveStallThreshold: Int = 3
    ) {
        self.tag = tag
        self.firstFrameTimeout = firstFrameTimeout
        self.freezeFrameThreshold = freezeFrameThreshold
        self.excessiveStallThreshold = excessiveStallThreshold
    }

    func resetForNewPlaybackSession() {
        playbackRequestedAt = 0
        playerReadyAt = 0
        firstFrameRenderedAt = 0
        lastFrameRenderedAt = 0
        lastKnownPlaybackTime = 0
        isPlaybackExpected = false
        isPlaying = false
        hasRenderedFirstFrame = false
        isBuffering = false
        isInFullscreenTransition = false
        isLayerAttached = false
        stallCount = 0
        layerAttachCount = 0
        fullscreenTransitionStartedAt = 0
        appDidEnterBackgroundAt = 0
        appWillEnterForegroundAt = 0
        lastPlaybackProgressedAt = 0
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        lastAccessLogSummary = ""
        recordedErrors.removeAll()
        errorOrder.removeAll()
        publishState("reset")
    }

    func onPlaybackRequested() {
        playbackRequestedAt = now()
        playerReadyAt = 0
        firstFrameRenderedAt = 0
        lastFrameRenderedAt = 0
        lastKnownPlaybackTime = 0
        lastPlaybackProgressedAt = 0
        isPlaybackExpected = true
        isPlaying = false
        isBuffering = false
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        publishState("playbackRequested")
    }

    func onPlayerReady() {
        if playerReadyAt == 0 {
            playerReadyAt = now()
        }
        publishState("playerReady")
    }

    func onPlaybackStarted() {
        isPlaying = true
        isPlaybackExpected = true
        publishState("playbackStarted")
    }

    func onPlaybackPaused() {
        isPlaying = false
        publishState("playbackPaused")
    }

    func onPlaybackCompleted() {
        isPlaying = false
        isPlaybackExpected = false
        isBuffering = false
        lastKnownPlaybackTime = 0
        lastPlaybackProgressedAt = 0
        lastFrameRenderedAt = 0
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        publishState("playbackCompleted")
    }

    func onFirstFrameRendered() {
        let timestamp = now()
        if firstFrameRenderedAt == 0 {
            firstFrameRenderedAt = timestamp
        }
        lastFrameRenderedAt = timestamp
        hasRenderedFirstFrame = true
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        publishState("firstFrameRendered")
    }

    func onFrameRendered() {
        lastFrameRenderedAt = now()
        if !hasRenderedFirstFrame {
            hasRenderedFirstFrame = true
            firstFrameRenderedAt = lastFrameRenderedAt
        }
        awaitingFullscreenRecovery = false
        awaitingBackgroundRecovery = false
        publishState("frameRendered")
    }

    func onTimeProgress(_ seconds: Double) {
        if seconds > lastKnownPlaybackTime {
            lastKnownPlaybackTime = seconds
            lastPlaybackProgressedAt = now()
        }
        publishState("timeProgress=\(String(format: "%.3f", seconds))")
    }

    func onBufferingStarted() {
        isBuffering = true
        publishState("bufferingStarted")
    }

    func onBufferingEnded() {
        isBuffering = false
        publishState("bufferingEnded")
    }

    func onStallDetected() {
        stallCount += 1
        if stallCount >= excessiveStallThreshold {
            recordError("EXCESSIVE_REBUFFERING")
        }
        publishState("stallDetected")
    }

    func onAudioMissing() {
        recordError("AUDIO_NOT_STARTED")
    }

    func onPlaybackNotStarted() {
        recordError("PLAYBACK_NOT_STARTED")
    }

    func onFullscreenTransitionStarted() {
        isInFullscreenTransition = true
        fullscreenTransitionStartedAt = now()
        awaitingFullscreenRecovery = isPlaybackExpected || isPlaying || hasRenderedFirstFrame
        publishState("fullscreenTransitionStarted")
    }

    func onFullscreenTransitionEnded() {
        isInFullscreenTransition = false
        if awaitingFullscreenRecovery {
            fullscreenTransitionStartedAt = now()
        }
        publishState("fullscreenTransitionEnded")
    }

    func onPlayerLayerAttached() {
        isLayerAttached = true
        layerAttachCount += 1
        if layerAttachCount >= 2 && !hasRenderedFirstFrame {
            recordError("DOUBLE_BLACK_SCREEN_RISK")
        }
        publishState("playerLayerAttached")
    }

    func onPlayerLayerDetached() {
        isLayerAttached = false
        publishState("playerLayerDetached")
    }

    func onAppDidEnterBackground() {
        appDidEnterBackgroundAt = now()
        awaitingBackgroundRecovery = isPlaybackExpected || isPlaying || hasRenderedFirstFrame
        isPlaying = false
        publishState("appDidEnterBackground")
    }

    func onAppWillEnterForeground() {
        appWillEnterForegroundAt = now()
        publishState("appWillEnterForeground")
    }

    func onAccessLogUpdate(summary: String) {
        lastAccessLogSummary = summary
        if summary.localizedCaseInsensitiveContains("error") ||
            ((_parseIntMetric(named: "stalls", from: summary) ?? 0) > 0) {
            onStallDetected()
        }
        publishState("accessLog=\(summary)")
    }

    func getErrors() -> [String] {
        errorOrder
    }

    func hasErrors() -> Bool {
        !errorOrder.isEmpty
    }

    func snapshot() -> [String: Any] {
        [
            "playbackRequestedAt": playbackRequestedAt,
            "playerReadyAt": playerReadyAt,
            "firstFrameRenderedAt": firstFrameRenderedAt,
            "lastFrameRenderedAt": lastFrameRenderedAt,
            "lastKnownPlaybackTime": lastKnownPlaybackTime,
            "isPlaybackExpected": isPlaybackExpected,
            "isPlaying": isPlaying,
            "hasRenderedFirstFrame": hasRenderedFirstFrame,
            "isBuffering": isBuffering,
            "isInFullscreenTransition": isInFullscreenTransition,
            "isLayerAttached": isLayerAttached,
            "stallCount": stallCount,
            "layerAttachCount": layerAttachCount,
            "fullscreenTransitionStartedAt": fullscreenTransitionStartedAt,
            "appDidEnterBackgroundAt": appDidEnterBackgroundAt,
            "appWillEnterForegroundAt": appWillEnterForegroundAt,
            "awaitingFullscreenRecovery": awaitingFullscreenRecovery,
            "awaitingBackgroundRecovery": awaitingBackgroundRecovery,
            "lastAccessLogSummary": lastAccessLogSummary,
            "errors": errorOrder,
        ]
    }

    func detectTimeoutAnomalies(referenceTime: CFTimeInterval = CACurrentMediaTime()) {
        let timestamp = referenceTime
        if playbackRequestedAt > 0 &&
            !hasRenderedFirstFrame &&
            timestamp - playbackRequestedAt > firstFrameTimeout {
            recordError("FIRST_FRAME_TIMEOUT")
        }
        if playerReadyAt > 0 &&
            !hasRenderedFirstFrame &&
            timestamp - playerReadyAt > firstFrameTimeout {
            recordError("READY_WITHOUT_FRAME")
        }
        if isPlaybackExpected &&
            !isPlaying &&
            !hasRenderedFirstFrame &&
            playbackRequestedAt > 0 &&
            timestamp - playbackRequestedAt > firstFrameTimeout {
            recordError("PLAYBACK_NOT_STARTED")
        }
        if awaitingFullscreenRecovery &&
            fullscreenTransitionStartedAt > 0 &&
            timestamp - fullscreenTransitionStartedAt > firstFrameTimeout &&
            isPlaybackExpected &&
            !hasFreshFrame(since: fullscreenTransitionStartedAt) {
            recordError("FULLSCREEN_INTERRUPTION")
        }
        if awaitingBackgroundRecovery &&
            appWillEnterForegroundAt > 0 &&
            timestamp - appWillEnterForegroundAt > firstFrameTimeout &&
            isPlaybackExpected &&
            !hasFreshFrame(since: appWillEnterForegroundAt) {
            recordError("BACKGROUND_RESUME_FAILURE")
        }
    }

    func detectVideoFreezeIfNeeded(
        referenceTime: CFTimeInterval = CACurrentMediaTime(),
        frameSilenceThreshold: CFTimeInterval? = nil
    ) {
        let timestamp = referenceTime
        let threshold = frameSilenceThreshold ?? freezeFrameThreshold
        guard isPlaybackExpected || isPlaying || lastKnownPlaybackTime > 0 else { return }
        guard hasRenderedFirstFrame else { return }
        guard lastFrameRenderedAt > 0 else { return }
        if timestamp - lastFrameRenderedAt > threshold {
            recordError("VIDEO_FREEZE")
        }
    }

    func onFullscreenInterruption() {
        recordError("FULLSCREEN_INTERRUPTION")
    }

    func onBackgroundResumeFailure() {
        recordError("BACKGROUND_RESUME_FAILURE")
    }

    func hasFreshFrame(since timestamp: CFTimeInterval) -> Bool {
        lastFrameRenderedAt >= timestamp && lastFrameRenderedAt > 0
    }

    func shouldFlagVideoFreeze(
        referenceTime: CFTimeInterval = CACurrentMediaTime(),
        frameSilenceThreshold: CFTimeInterval? = nil
    ) -> Bool {
        let threshold = frameSilenceThreshold ?? freezeFrameThreshold
        guard hasRenderedFirstFrame else { return false }
        guard isPlaybackExpected || isPlaying || lastPlaybackProgressedAt > 0 else { return false }
        guard lastFrameRenderedAt > 0 else { return false }
        return referenceTime - lastFrameRenderedAt > threshold
    }

    private func recordError(_ code: String) {
        guard !recordedErrors.contains(code) else { return }
        recordedErrors.insert(code)
        errorOrder.append(code)
        log("error=\(code)")
        publishState("error=\(code)")
    }

    private func _parseIntMetric(named metric: String, from summary: String) -> Int? {
        guard let range = summary.range(of: "\(metric)=") else { return nil }
        let suffix = summary[range.upperBound...]
        let digits = suffix.prefix { $0.isNumber }
        guard !digits.isEmpty else { return nil }
        return Int(digits)
    }

    private func publishState(_ event: String) {
        log(event)
        stateListener?(self)
    }

    private func log(_ message: String) {
        print("[\(tag)] \(message)")
    }

    private func now() -> CFTimeInterval {
        CACurrentMediaTime()
    }
}
